crates/persistence_api/src/eav/eavi.rs

```diff
@@ -0,0 +1,493 @@
+        let value =
+            ExampleAddressableContent::try_from_content(&JsonString::from(RawString::from("blue")))
+                .unwrap();
+
+        EavTestSuite::test_round_trip(eav_storage, entity, attribute, value)
+    }
+
+    #[test]
+    fn example_eav_one_to_many() {
+        EavTestSuite::test_one_to_many::<
+            ExampleAddressableContent,
+            ExampleEntityAttributeValueStorage<ExampleAttribute>,
+        >(test_eav_storage());
+    }
+
+    #[test]
+    fn example_eav_many_to_one() {
+        EavTestSuite::test_many_to_one::<
+            ExampleAddressableContent,
+            ExampleEntityAttributeValueStorage<ExampleAttribute>,
+        >(test_eav_storage());
+    }
+
+    #[test]
+    fn example_eav_range() {
+        EavTestSuite::test_range::<
+            ExampleAddressableContent,
+            ExampleEntityAttributeValueStorage<ExampleAttribute>,
+        >(test_eav_storage());
+    }
+
+    #[test]
+    fn example_eav_prefixes() {
+        EavTestSuite::test_multiple_attributes::<
+            ExampleAddressableContent,
+            ExampleEntityAttributeValueStorage<ExampleAttribute>,
+        >(test_eav_storage(), {
+            let mut attrs: Vec<ExampleAttribute> = vec!["a_", "b_", "c_", "d_"]
+                .into_iter()
+                .map(|p| ExampleAttribute::WithPayload(p.to_string() + "one_to_many"))
+                .collect();
+            attrs.push(ExampleAttribute::WithoutPayload);
+            attrs
+        });
+    }
+
+    #[test]
+    /// show AddressableContent implementation
+    fn addressable_content_test() {
+        // from_content()
+        AddressableContentTestSuite::addressable_content_trait_test::<
+            EntityAttributeValueIndex<ExampleAttribute>,
+        >(test_eav_content(), test_eav(), test_eav_address());
+    }
+
+    #[test]
+    /// show CAS round trip
+    fn cas_round_trip_test() {
+        let addressable_contents = vec![test_eav()];
+        AddressableContentTestSuite::addressable_content_round_trip::<
+            EntityAttributeValueIndex<ExampleAttribute>,
+            ExampleContentAddressableStorage,
+        >(addressable_contents, test_content_addressable_storage());
+    }
+
+    #[test]
+    fn validate_attribute_paths() {
+        assert!(EntityAttributeValueIndex::new(
+            &test_eav_entity().address(),
+            &ExampleAttribute::WithPayload("abc".into()),
+            &test_eav_entity().address()
+        )
+        .is_ok());
+        assert!(EntityAttributeValueIndex::new(
+            &test_eav_entity().address(),
+            &ExampleAttribute::WithPayload("abc123".into()),
+            &test_eav_entity().address()
+        )
+        .is_ok());
+        assert!(EntityAttributeValueIndex::new(
+            &test_eav_entity().address(),
+            &ExampleAttribute::WithPayload("123".into()),
+            &test_eav_entity().address()
+        )
+        .is_ok());
+    }
+
+}
```

Comment:
I'm interested in the decision to make Attribute a trait, could you walk me through that?

Reply:
The attribute in holochain code was very specific to managing crud and links. The generalized form using an  attribute trait allows for other representations.